### PR TITLE
[RFC] React Primitives

### PR DESCRIPTION
**Pull request by mxstbr**
**Status**
- [ ] WIP
- [x] Ready for review
- [ ] Needs testing

This commit introduces [react-primitives](https://github.com/lelandrichardson/react-primitives) to the mobile app codebase.

react-primitives exposes a limited set of ReactNative-alike APIs (among them `<View>`, `<Text>` and `<Touchable>`) and makes them work across many React render targets.

We decided against it when we started building the mobile app because it wasn't maintained at the time and didn't work with React v16, a deal breaker for us. Nowadays it's maintained again, (thanks to the amazing @MathieuDutour) so both of these blockers have been resolved.

----

This PR adds the react-primitives dependency and reworks the mobile `<Text>` component to be implemented with primitives. **This allows the `<Text>` component to be rendered in the native app, in the web app, and even in Sketch and ReactVR—it's now truly universal!**

As you can see in the changes, styled-components also has built-in support for `react-primitives` by importing from `styled-components/primitives` instead of `/native`.

While there's always parts of the apps that we'll have to build with their native APIs, we should be able to build a large portion of our mobile component library based on primitives, allowing us to potentially reuse a lot of the work across our platforms in the future.

It shouldn't be too much of a difference compared to building components with vanilla ReactNative, although you might have to be a bit creative to stay within the constraints set forth by react-primitives. (the limited API it exposes) We'll also never be able to build 100% of our mobile app with it, as parts of the app do require native APIs that simply don't exist on other platforms. (think push notifications, sharing, etc.)

Should we decide to use `react-primitives` here's how I think it'd make the most sense to approach it:

1. Use it for all new, reusable components in the mobile app
1. Slowly move over the existing components of the mobile app as we touch them
1. Start using universal components built for new features across the mobile and web apps
1. Eventually in the far future maybe move the web app over to the new, resuable component libarary

Let's discuss about it! @brianlovin @uberbryn any thoughts?

/cc @jongold who I'm sure has something good to say about primitives :p

### CODE REVIEW

File: mobile/components/Text/index.js [hunk 1]
```diff
@@ -45,8 +45,8 @@ const monospaceFont = Platform.OS === 'android' ? 'monospace' : 'Menlo';
 const Text: ComponentType<Props> = styled.Text`
   font-weight: ${props =>
     props.weight ? `${getWeightFromType(props.weight)}` : '400'};
-  font-size: ${props => getTypeConfigFromType(props.type).size}px;
-  line-height: ${props => getTypeConfigFromType(props.type).leading}px;
+  font-size: ${props => getTypeConfigFromType(props.type).size};
+  line-height: ${props => getTypeConfigFromType(props.type).leading};
```
Comment (mxstbr) [inline mobile/components/Text/index.js:49]: As you can see, the biggest different between vanilla RN and react-primitives (apart from the minimal API) is that styled-components doesn't do any fancy "make it look like web CSS" tricks—instead it just takes the string and turns it into an object, so `px` is not required anymore.

**Comment by mxstbr** [comment]
This allows one to render the `<Text>` component to web, but also to Sketch! 😍

```JS
import * as React from 'react';
import { ThemeProvider } from 'styled-components';
import { render, Artboard, View } from 'react-sketchapp';
import Text from 'mobile/components/Text';
import theme from 'mobile/components/theme';

const Document = () => (
  <Artboard
    name="Spectrum"
    style={{
      flexDirection: 'row',
      flexWrap: 'wrap',
      width: (96 + 8) * 4,
    }}
  >
    <ThemeProvider theme={theme}>
      <View>
        <Text type="largeTitle">Large Title</Text>
        <Text type="title1">Title 1</Text>
        <Text type="title2">Title 2</Text>
        <Text type="title3">Title 3</Text>
        <Text type="headline">Headline</Text>
        <Text type="body">Body Text</Text>
        <Text type="body" weight="bold">Bold Body Text</Text>
        <Text type="body" italic>Italic Body Text</Text>
      </View>
    </ThemeProvider>
  </Artboard>
);

export default () => {
  render(<Document />, context.document.currentPage());
};
```

screen shot 2018-06-21 at 14 47 34

**Comment by brianlovin** [comment]
I'm not opposed, just pretty concerned that we are going to run into bugs and quirks with this setup that slow us from shipping the mobile app. We won't know until we try, which makes the decision hard.

I'll leave it up to you and will learn the new thing if you decide to go with this proposal.

**Review by brianlovin — APPROVED** [review]

**Comment by ryota-murakami** [comment]
I think React Primitives is ideal solution for closer to "Write Once React Application, Run Cross-Platform".
This meant we don't need build developer team for each platforms and has possibility reduce codebase fragmentation.

I was consider whether adopt React Primitives in my first PR https://github.com/withspectrum/spectrum/pull/2880😜

But i didn't know whether does it work existing web css correctly on ReactNative.
For Example,

- pseudo selector
- media query
- difference flexbox specification(web flexbox default is `flex-direction: row`, but ReactNative flexbox default is `flex-direction: column`)
- and other invalid css rule/property in ReactNative...

That concern is... which is no worry?

**Comment by mxstbr** [comment]
We decided against react-primitives for now, mainly due to the limited manpower we have but also because it'd make SSR a lot harder. Next time!

**Pull request closed without merging**